app/jobs/base.rb

```diff
@@ -16,35 +16,87 @@ def self.num_email_retry_jobs
   end
 
   class Base
+    class JobInstrumenter
+      def initialize(job_class:, opts:, db:)
+        return unless enabled?
+        @data = {}
+
+        @data["hostname"] = `hostname`.strip # Hostname
+        @data["pid"] = Process.pid # Pid
+        @data["database"] = db # DB name - multisite db name it ran on
+        @data["job_name"] = job_class.name # Job Name - eg: Jobs::AboutStats
+        @data["job_type"] = job_class.try(:scheduled?) ? "scheduled" : "regular" # Job Type - either s for scheduled or r for regular
+        @data["opts"] = opts # Params - json encoded params for the job
+
+        @data["status"] = 'pending'
+        @start_timestamp = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+        self.class.ensure_interval_logging!
+        @@active_jobs ||= []
+        @@active_jobs << self
+
+        MethodProfiler.ensure_discourse_instrumentation!
+        MethodProfiler.start
+      end
 
-    class Instrumenter
+      def stop(exception:)
+        return unless enabled?
 
-      def self.stats
-        Thread.current[:db_stats] ||= Stats.new
-      end
+        profile = MethodProfiler.stop
+
+        @@active_jobs.delete(self)
 
-      class Stats
-        attr_accessor :query_count, :duration_ms
+        @data["duration"] = profile[:total_duration] # Duration - length in seconds it took to run
+        @data["sql_duration"] = profile.dig(:sql, :duration) || 0 # Sql Duration (s)
+        @data["sql_calls"] = profile.dig(:sql, :calls) || 0 # Sql Statements - how many statements ran
+        @data["redis_duration"] = profile.dig(:redis, :duration) || 0 # Redis Duration (s)
+        @data["redis_calls"] = profile.dig(:redis, :calls) || 0 # Redis commands
+        @data["net_duration"] = profile.dig(:net, :duration) || 0 # Redis Duration (s)
+        @data["net_calls"] = profile.dig(:net, :calls) || 0 # Redis commands
 
-        def initialize
-          @query_count = 0
-          @duration_ms = 0
+        if exception.present?
+          @data["exception"] = exception # Exception - if job fails a json encoded exception
+          @data["status"] = 'failed'
+        else
+          @data["status"] = 'success' # Status - fail, success, pending
         end
+
+        write_to_log
       end
 
-      def call(name, start, finish, message_id, values)
-        stats = Instrumenter.stats
-        stats.query_count += 1
-        stats.duration_ms += (((finish - start).to_f) * 1000).to_i
+      def logger
+        @@logger ||= Logger.new("#{Rails.root}/log/sidekiq.log")
```

Comment:
I think we want an async logger here so there is no chance logging blocks sidekiq, I think active support has one of those.

Reply:
I don't think ActiveSupport has anything. It [used to have](https://github.com/rails/rails/commit/572c3d517899524c2a7c4c84ad9646660168d4cd#diff-03c8fca389d378ba1a40492d96b5fdc5) `BufferedLogger`, but that was removed a while ago.

I've implemented a queue, with logging in a separate thread (3b84d98), but I'm not sure if the performance benefit is worth the extra code, especially bearing in mind that this entire thing is opt-in? Let me know what you think.